### PR TITLE
feat(NewMessageUploadEditor) - caption to file share

### LINKED ISSUE #10414: Frontend "File caption"
- Overview https://github.com/nextcloud/spreed/issues/5354

---

- Draft at https://github.com/nextcloud/spreed/pull/8949

### PR DESCRIPTION
**Pull request by Antreesy**
### ☑️ Resolves

* Fix #10414

### 🖼️ Screenshots / Screencasts

🏚️ Before | 🏡 After
---|---
image | Screenshot from 2023-10-18 17-22-50
image | Screenshot from 2023-10-18 17-29-01

### 🚧 Tasks

- [ ] To check
  - [ ] User mentions don't render for me in NewMessage (known bug)
  - [ ] Caught it once, the order of shared files was different, because they are now being loaded in parallel
- [ ] Follow-up
  - [ ] Support multiple file shares (requires a lot of rework in frontend/backend) - draft at #8949

### 🏁 Checklist

- [x] 🌏 Tested with Chrome, Firefox and Safari or should not be risky to browser differences
- [x] 🖌️ Design was reviewed, approved or inspired by the design team
- [x] ⛑️ Tests are included or not possible
- [ ] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required

### CODE REVIEW

File: src/components/MessagesList/MessagesGroup/Message/Message.vue [hunk 1]
```diff
@@ -502,6 +502,14 @@ export default {
 			return !this.isLastMessage && this.id === this.$store.getters.getVisualLastReadMessageId(this.token)
 		},
 
+		renderedMessage() {
+			if (this.messageParameters?.file && this.message !== '{file}') {
+				return `{file} ${this.message}`
```
Comment (nickvergessen) [inline src/components/MessagesList/MessagesGroup/Message/Message.vue:507]: space? Or should be a new line or something?

Comment (Antreesy) [reply src/components/MessagesList/MessagesGroup/Message/Message.vue:507]: new line `\n\n` would be better, as it puts files and caption in different paragraphs
Although it creates a gap under the image, and doesn't look good with one-liners, we could stylize them later

File: src/components/NewMessage/NewMessage.vue [hunk 2]
```diff
@@ -466,6 +481,12 @@ export default {
 				}
 			}
 
+			if (this.upload) {
+				this.$emit('sent', this.text)
+				this.$store.dispatch('setCurrentMessageInput', { token: this.token, text: '' })
```
Comment (DorraJaouad) [inline src/components/NewMessage/NewMessage.vue:486]: There should be a smooth scrolling to the bottom after upload is **completely** finished but don't know where to emit the event

File: src/store/fileUploadStore.js [hunk 3]
```diff
@@ -282,31 +270,38 @@ const actions = {
...
+		// If caption is provided, attach to the last temporary message
+		const lastIndex = getters.getUploadsArray(uploadId).at(-1).at(0)
+		for (const [index, uploadedFile] of getters.getUploadsArray(uploadId)) {
 			// mark all files as uploading
 			commit('markFileAsUploading', { uploadId, index })
 			// Store the previously created temporary message
-			const temporaryMessage = state.uploads[uploadId].files[index].temporaryMessage
+			const temporaryMessage = {
+				...uploadedFile.temporaryMessage,
+				message: index === lastIndex ? caption : '{file}',
```
Comment (DorraJaouad) [inline src/store/fileUploadStore.js:293]: Maybe it's better to store the rendered message format here and use it in `Message` . After all, the message contains a file and usually there is {file} to indicate its presence in `Message.message`

File: src/components/MessagesList/MessagesGroup/Message/Message.vue [hunk 4]
```diff
@@ -568,7 +576,7 @@ export default {
 			let match
 			let emojiStrings = ''
 			let emojiCount = 0
-			const trimmedMessage = this.message.trim()
+			const trimmedMessage = this.renderedMessage.trim()
```
[the remark this answers is not in the document]
Comment (Antreesy) [reply src/components/MessagesList/MessagesGroup/Message/Message.vue:580]: Do you mean `containsCodeBlocks()`? then code block will be passed as caption, so `this.message` should contains it. Although we should test that, nice find

Or `<MessageTranslateDialog />`? Then we shouldn't translate {file} part, so passing `this.message` is also fine

File: src/components/NewMessage/NewMessage.vue (hunk 2 above)
Comment (Antreesy) [reply src/components/NewMessage/NewMessage.vue:486]: Maybe we could use a preview from temp message behind loading spinner, so height will remain. But should check, how will that work for failed uploads

File: src/store/fileUploadStore.js (hunk 3 above)
Comment (Antreesy) [reply src/store/fileUploadStore.js:293]: We got either 'caption' or '{file}' in the server response, so temp message should follow the same logic here, IMO

File: src/components/MessagesList/MessagesGroup/Message/Message.vue (hunk 4 above)
[the remark this answers is not in the document]
Comment (DorraJaouad) [reply src/components/MessagesList/MessagesGroup/Message/Message.vue:580]: yes, I meant `containsCodeBlocks()`

File: src/components/NewMessage/NewMessage.vue (hunk 2 above)
Comment (Antreesy) [reply src/components/NewMessage/NewMessage.vue:486]: I researched a bit: this has nothing to do with this PR, so I'd raise a follow-up here (or there is probably an issue for this already)

Pushed a new commit to use local images (from blob) and reduce server load and list flickering

**Pull request merged**